File: haskell-server/src/EagerWords/Common/Domain/PiecePoint.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE StandaloneDeriving #-}

module EagerWords.Common.Domain.PiecePoint (
    PiecePoint(..)
  , pointLetter
)
where

import GHC.Generics (Generic)
import Control.DeepSeq (NFData)
import Data.Aeson (FromJSON, ToJSON)
import EagerWords.Common.Domain.Piece
import qualified EagerWords.Common.Domain.Piece as Piece
import Bolour.Plane.Domain.Point (Point)

-- | A piece located at a point on the plane.
data PiecePoint = PiecePoint {
    piece :: Piece    -- ^ The piece.
  , point :: Point    -- ^ The position of the piece on the plane.
} deriving (Eq, Show, Generic, NFData)

instance FromJSON PiecePoint
instance ToJSON PiecePoint

pointLetter :: PiecePoint -> Char
pointLetter PiecePoint {piece} = Piece.value piece

File: scala-server/src/main/scala/com/bolour/eagerwords/common/domain/PiecePoint.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords.com/blob/master/LICENSE.md
 */
package com.bolour.eagerwords.common.domain

import com.bolour.grid.domain.Point

/**
  * Combination of a piece and its location on the board.
  */
case class PiecePoint(piece: Piece, point: Point)
